lib/internal/url.js

```diff
@@ -932,6 +935,79 @@ function urlToOptions(url) {
   return options;
 }
 
+function getPathFromURLWin32(url) {
+  var hostname = url.hostname;
+  var pathname = url.pathname;
+  if (hostname !== '') {
+    // If hostname is set, then we have a UNC path
+    return `//${hostname}${pathname}`;
+  } else {
+    // Otherwise, it's a local path that requires a drive letter
+    var letter = pathname.codePointAt(1) | 0x20;
+    var sep = pathname[2];
+    if (letter < 96 || letter > 122 ||
+        (sep !== ':' && sep !== '|')) {
+      return new Error('File URLs must specify absolute paths');
+    }
+    pathname = pathname.slice(1);
+    for (var n = 0; n < pathname.length; n++) {
+      if (pathname[n] === '%') {
+        var third = pathname.codePointAt(n + 2) | 0x20;
+        if ((pathname[n + 1] === '2' && third === 102) ||
+            (pathname[n + 1] === '5' && third === 99)) {
+          return new Error('Path must not include encoded \\ or / characters');
+        }
+      }
+    }
+    return decodeURIComponent(pathname);
+  }
+}
+
+function getPathFromURLOther(url) {
+  if (url.hostname !== '') {
+    return new Error(
+      `File URLs on ${os.platform()} must use hostname localhost`);
+  }
+  var pathname = url.pathname;
+  for (var n = 0; n < pathname.length; n++) {
+    if (pathname[n] === '%') {
+      var third = pathname.codePointAt(n + 2) | 0x20;
+      if (pathname[n + 1] === '2' && third === 102) {
+        return new Error('Path must not include encoded / characters');
+      }
+    }
+  }
+  return decodeURIComponent(pathname);
+}
+
+function isFileUrl(path) {
+  if (typeof path !== 'string') return false;
+  if (path.length < 7) return false;
+  if ((path.codePointAt(0) | 0x20) !== 102) return false; // f F
+  if ((path.codePointAt(1) | 0x20) !== 105) return false; // i I
+  if ((path.codePointAt(2) | 0x20) !== 108) return false; // l L
+  if ((path.codePointAt(3) | 0x20) !== 101) return false; // e E
+  if (path[4] !== ':' || path[5] !== '/' || path[6] !== '/') {
+    return false;
+  }
+  return true;
+}
+
+function getPathFromURL(path) {
+  var url;
+  if (isFileUrl(path)) {
+    url = new URL(path);
```

Comment:
Are we sure this is safe? It feels like the kind of thing that could mess up any currently existing checks that userland applications run on the pathnames… Also, technically `file:///asdf` is a perfectly valid relative path on POSIX, so that alone makes this a breaking change.
I’d feel a lot better if @nodejs/security could give this a look.

And if we make decisions based on `isFileUrl(path)`, we should probably make that function accessible from userland, too.